pkg/pubsub/pubsub.go

```diff
@@ -0,0 +1,106 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pubsub
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+//PubSub holds publishers and subscribers
+type PubSub struct {
+	subs  []chan interface{}
+	pub   chan interface{}
+	mutex sync.Mutex
+	count int64
+	buf   []interface{}
+}
+
+var (
+	bufferCount = 1
+)
+
+// process item to subscribers.
+func (ps *PubSub) process() {
+	for item := range ps.pub {
+		ps.mutex.Lock()
+		for _, sub := range ps.subs {
+			go func(s chan interface{}) {
+				s <- item
+			}(sub)
+		}
+		ps.mutex.Unlock()
+	}
+}
+
+// Publish message to pubsub system
+func (ps *PubSub) Publish(item interface{}) {
+	count := int(atomic.AddInt64(&ps.count, 1))
+	if count > bufferCount {
+		// reset count
+		atomic.StoreInt64(&ps.count, 0)
+		for _, msg := range ps.buf {
+			ps.pub <- msg
+		}
+	}
+	if count <= bufferCount {
+		idx := count % bufferCount
+		ps.buf[idx] = item
+	}
```

Comment:
buffered chans should be used instead of using custom `[] ps.buf`
Also a go routine should be used to send to the buffered channel so that `Publish()` returns immediately

Reply:
you can create buffered chans for publisher chan and subcriber chan